calico/operations/ebpf/troubleshoot-ebpf.mdx

```diff
@@ -202,9 +242,89 @@ The parts of the log are explained below:
   - A host endpoint ingress program is executed on the path from external node to the host.
   - A host endpoint egress program is executed on the path from host to external host.
 
+- you may also see `ens192---X` which indicates an XDP program. $[prodname] uses XDP programs to implement `doNotTrack` policies on host devices only.
+
 - `Final result=ALLOW (-1). Program execution time: 7366ns` is the message. In this case, logging the final result of
   the program. Note that the timestamp is massively distorted by the time spent logging.
 
+## Debugging policy issues
+
+If you suspect that $[prodname] is dropping packets due to policy, you can use the
+`calico-node -bpf` tool to dump the policy that is attached to a specific interface.
+
+```
+kubectl exec -n calico-system calico-node-abcdef -- calico-node -bpf policy dump <interface> <hook>
+
+where `<interface>` is the name of the interface and `<hook>` is either `ingress`, `egress`, `xdp` or `all`.
+```
+
+Dump of an ingress policy. Note that ingress policy for a pod is attached to the egress hook of the host-side of the caliX veth pair.
+
+```
+IfaceName: calic31b4f7fc58
+Hook: tc egress
+Error:
+Policy Info:
+// Start of tier default
+// Start of policy default/knp.default.allow-nginx-from-ubuntu
+// Start of rule action:"allow"  protocol:{name:"tcp"}  dst_ports:{first:80  last:80}  src_ip_set_ids:"s:nzE8vwTu69FSscx2FDKjb20D9dZxEyVxsWFqwA"  original_src_selector:"projectcalico.org/orchestrator == 'k8s' && app == 'ubuntu-client'"  rule_id:"29vMYcPWr7reSxxN"
+// IPSets src_ip_set_ids:<0x303904ae5eae5418>
+// count = 9
+// End of rule 29vMYcPWr7reSxxN
+// End of policy default/knp.default.allow-nginx-from-ubuntu
+// End of tier default: deny
+// Start of rule action:"allow"  rule_id:"aBMQCbsUMESPKGRp"
+// count = 0
+// End of rule aBMQCbsUMESPKGRp
+```
+
+Rules that use selectors refer to IP sets. You can dump the contents of an IP set using the `ipsets` command and you can check whether the IP set contains the expected members:
+
+You can see how many packets have matched each rule. In this example, 9 packets have matched rule `29vMYcPWr7reSxxN` and 0 packets have matched rule `aBMQCbsUMESPKGRp`.
+
+Adding `--asm` will show the eBPF assembly code for the program as well.
+
+```
+kubectl exec -n calico-system calico-node-abcdef -- calico-node -bpf ipsets dump
+
+IP set 0x303904ae5eae5418
+   192.168.58.5/32
+
+IP set 0xffef9f925a8a4ca4
+   192.168.129.66/32
+   192.168.129.68/32
+```
+
+## Debugging calico-node not ready
+
+If you notice that a `calico-node` pod is not ready, check its logs for errors. The most likely reason for `calico-node` not being ready in eBPF mode is
+that $[prodname] is not able to update a program attached to an interface. Look for the following type of warning:
+
+```
+2025-09-22 22:39:59.801 [WARNING][10374] felix/bpf_ep_mgr.go 2107: Failed to apply policy to endpoint, leaving it dirty error=<error message>
```

Comment:
Better to use an actual error message here, since you're looking at a specific warning.

Alternatively, you can cut off the sample earlier.
```
2025-09-22 22:39:59.801 [WARNING][10374] felix/bpf_ep_mgr.go 2107: Failed to apply policy to endpoint, leaving it dirty